src/understand-tidb/mvcc-garbage-collection.md

```diff
@@ -1 +1,280 @@
+
+```go
+// Start starts the worker.
+func (w *GCWorker) Start() {
+	var ctx context.Context
+	ctx, w.cancel = context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go w.start(ctx, &wg)
+	wg.Wait() // Wait create session finish in worker, some test code depend on this to avoid race.
+}
+```
+
+There is a [ticker](https://github.com/pingcap/tidb/blob/v5.2.1/store/gcworker/gc_worker.go#L214-L229) in `GCWorker` that will try running the GC task every minute. We'll talk about how it works in the following sections.
+
+### GC Leader
+
+In TiDB's design, there is only one GC leader which can trigger GC in the cluster, so there is an election and lease for the proposal of the leader. `GCWorker` check if it's the leader before starting a real GC job.
+
+```go
+func (w *GCWorker) tick(ctx context.Context) {
+	isLeader, err := w.checkLeader()
+	...
+	if isLeader {
+		err = w.leaderTick(ctx)
+		...
+	} else {
+		// Config metrics should always be updated by leader, set them to 0 when current instance is not leader.
+		metrics.GCConfigGauge.WithLabelValues(gcRunIntervalKey).Set(0)
+		metrics.GCConfigGauge.WithLabelValues(gcLifeTimeKey).Set(0)
+	}
+}
+```
+
+Luckily, TiDB is built on a durable and high-available storage layer, which makes the election quite easy by the following mechanism.
+
+![gc-leader-election.png](../img/gc-leader-election.png)
+
+This is the flowchart of leader election which is implemented in [`checkLeader`](https://github.com/pingcap/tidb/blob/v5.2.1/store/gcworker/gc_worker.go#L1669) function, if the result is leader, we then trying to tick a GC.
+
+### GC Prepare
+
+When a GC workflow is triggered, we are facing another issue, find out the data that can be cleaned up. Since TiDB follows SI isolation level, all reads performed in a transaction should get the result from the same timestamp, if not so, there will be consistency and repeatable read issues. Besides, we do not hope that the garbage data is cleanup once after it's not used, in some misoperations, recovery is important. That means before starting a GC round, there are [many checks](https://github.com/pingcap/tidb/blob/v5.2.1/store/gcworker/gc_worker.go#L362-L393).
+
+```go
+func (w *GCWorker) checkPrepare(ctx context.Context) (bool, uint64, error) {
+	enable, err := w.checkGCEnable()
+	...
+	now, err := w.getOracleTime()
+	...
+	ok, err := w.checkGCInterval(now)
+	...
+	newSafePoint, newSafePointValue, err := w.calcNewSafePoint(ctx, now)
+	...
+	err = w.saveTime(gcLastRunTimeKey, now)
+	...
+	err = w.saveTime(gcSafePointKey, *newSafePoint)
+	...
+	return true, newSafePointValue, nil
+}
+```
+
+Check GC is enabled first, GC is enabled by default, but it's allowed to be turned off.
+
+In TiDB, there is a variable named `tikv_gc_run_interval` which controls the frequency of GC. It's the min interval between 2 GC rounds, which guarantees that data should live more than the interval time, 10min by default.
```

Comment:
```suggestion
In TiDB, there is a variable named `tikv_gc_run_interval` which controls the frequency of GC. It's the min interval between two GC rounds, which guarantees that data should live more than the interval time, 10min by default.
```

Reply:
How does it guarantee that data should live more than the interval time?

If a data becomes garbage just at the moment before a GC trigger, it will immediately participant the GC workflow. Unless there is a checker that it escapes for the first round always.